proposals/LocalRewrites.md

```diff
@@ -0,0 +1,124 @@
+# Local Rewrites Proposal
+
+## The Issue
+
+Suppose we are working with the following Liquid Haskell code:
+
+```haskell
+{-@ LIQUID "--ple" @-}
+{-@ LIQUID "--reflection" @-}
+
+module Example where
+
+import Language.Haskell.Liquid.ProofCombinators
+import Prelude hiding (id)
+
+{-@ reflect id @-}
+id :: Int -> Int
+id x = x
+
+data Term where
+    {-@ MkId :: Prop (Term id) @-}
+    MkId :: Term
+data TERM = Term (Int -> Int)
+```
+
+Parametrizing a type with some value is a common pattern in dependently typed
+programming languages.
+
+We would like to prove statements that depend on these parameters.
+
+```haskell
+{-@ doSomething :: f:_ -> Prop (Term f) -> {f = id} @-}
+doSomething :: (Int -> Int) -> Term -> Proof
+doSomething _ MkId = trivial
+```
+
+The above passes verification because the SMT solver can easily infer that if
+the term was constructed using the `MkId` constructor, then `f = id`.
+
+However, in many cases, this is not enough because there are contexts where the
+SMT solver's knowledge is insufficient. Letting PLE unfold `f` can lead to
+verification. Consider the following example:
+
+```haskell
+{-@ doSomething' :: f:_ -> Prop (Term f) -> {f 42 = 42} @-}
+doSomething' :: (Int -> Int) -> Term -> Proof
+doSomething' _ MkId = trivial
+```
+
+Since for the SMT solver, `f` is still an opaque term (even though it knows that
+`f = id`), and PLE doesn't have any equation that states `f = id`, the
+verification fails.
+
+This issue could be solved by recognizing that `f` must be equal to `id` when
+case-splitting and by informing PLE to expand `f -> id`. However, this expansion
+must be local. If we add another constructor to `Term`, `f = id` will not hold
+universally.
+
+```haskell
+data Term where
+    {-@ MkId :: Prop (Term id) @-}
+    MkId :: Term
+    {-@ MkAny :: f:_ -> Prop (Term f) @-}
+    MkAny :: (Int -> Int) -> Term
+data TERM = Term (Int -> Int)
+```
+
+Currently, the way Liquid Haskell communicates with Liquid Fixpoint, and by
+extension PLE, is through the `define` directive. For example, for `id`, it
+would generate the following definition:
+
+```fq
+define Example.id (x : Int) = {(x)}
+```
+
+The issue is that this mechanism is global—we cannot specify that the rewrite
+should occur only in specific subconstraints.
+
+## Possible Solutions
+
+### Solution 1: Modification of Various Frontends (Liquid Haskell)
+
+This proposal suggests the following steps:
+
+- Introduce a mechanism to provide local definitions in the `.fq` file format.
+  For example: `define n f = {(Example.id)}`, where `n` is the binding ID that
```

Comment:
I'm just noticing that if the new define form allowed parameters, as the old form does, this mechanism could be reused to unfold local recursive definitions as well.

Reply:
You're right but first we need to allow lambdas in reflected function as GHC is generating lambda when using where clauses to define functions